core/src/main/java/org/infinispan/persistence/sifs/IndexNode.java

```diff
@@ -1115,46 +1175,162 @@ public String toString() {
 
    Flowable<EntryRecord> publish(IntSet segments, boolean loadValues) {
       long currentTime = segment.getTimeService().wallClockTime();
-      // TODO: this is not non blocking
-      return Flowable.create(emitter -> {
-         try {
-            recursiveNode(this, segment, segments, emitter, loadValues, currentTime);
-         } catch (Throwable t) {
-            emitter.onError(t);
+
+      int cacheSegmentSize = segments.size();
+      if (cacheSegmentSize == 0) {
+         return Flowable.empty();
+      }
+
+      // Needs defer as we mutate the deque so publisher can be subscribed to multiple times
+      return Flowable.defer(() -> {
+         // First sort all the segments by their unsigned numeric byte[] values.
```

Comment:
We could avoid the sort step (or at least do the sorting by copying the segments into an BitSet instead of creating a  byte[] for each) if we used a big-endian 2-byte encoding for the segment.

Reply:
I don't think we can avoid the byte[] allocation as we need the segment prefix to find where in the binary tree to go first.

Reply:
I believe `getIterationPoint()` is only ever called with a segment prefix, so it could read the first 2 bytes of the prefix as an int to compare them with the parameter (also passed as an int).

Even the binary search could be customized to look at only one byte or two bytes passed as an int, but I expect the situations where the prefix has 0 or 1 bytes aren't very common, so probably not worth optimizing.